addon/resolve-attribute-util.js

```diff
@@ -12,19 +12,20 @@ import {
 } from './utils/resolve';
 
 function resolveReference(store, reference) {
+  let { id } = reference;
   if (reference.type === null) {
     // for schemas with a global id-space but multiple types, schemas may
     // report a type of null
-    let internalModel = store._globalM3Cache[reference.id];
+    let internalModel = store._globalM3Cache[id];
     return internalModel ? internalModel.getRecord() : null;
   } else {
     // respect the user schema's type if provided
-    return store.peekRecord(reference.type, reference.id);
+    return id !== null && id !== undefined ? store.peekRecord(reference.type, reference.id) : null;
   }
 }
 
 function resolveReferenceOrReferences(store, model, key, value, reference) {
-  if (Array.isArray(value) || Array.isArray(reference)) {
+  if (Array.isArray(reference)) {
```

Comment:
what's the motivation here? Isn't that at odds with the concept of "always sync"?

Reply:
Broadly it's to always prioritize user signal over convention or heuristics.

Specifically if you have a payload like

```js
{
  foreignKeys: ['urn:book:1'],
  resolvedForeignKeys: null // due to resolution failure; perhaps b/c a user deleted the record or something
}
```

Whether this should be an empty reference array or `null` should be up to the user